hype/fees.md

```diff
@@ -0,0 +1,112 @@
+    pub blobs: Vec<Blob>,
+}
+```
+
+When a new transaction is sent to the node, it runs the `hyfi` & `hydentity` smart contracts code with blobs built from `Fees` attributes.
+
+➡️  For now we check that `fees.identity_contract == "hydentity" && fees.fee_contract == "hyfi"`, but later we could consider whitelists.
+
+If the execution succeeds **and** resulting balance of `payer` is greater that `base_amount` (Cf next ⚠️ ), it can be sent to mempool for data dissemination. A proof will be generated later.
+
+⚠️  At this stage, the node knows it will be able to generate a proof for this blob and gather the fees. Modulo there is no other transaction 
+in an other batch that would be included before
+    - if such a case happens, an `identity` can have a negative balance
+    - to limit the "negative balance" issue, we can consider that an `identity` needs to always have at least `base_amount` of token. If the payer has a balance under this amount the fees won't 
+be accepted, but if the payers got multiple transactions in different lanes that ended to pay more fees than allowed, the payer won't get a negative balance. The payer won't be able to pay for fees
+until it gains some more tokens.
+    - It does not 100% delete the "negative balance issue", unless this `base_amount` is high enough... This is an hyper parameter to find.
+
+❓ Should each node do this verification when receiving the tx in data dissemination ? If we don't, the network might accept transactions with no fees if a node "don't care" about fees in its data lane.
+- Could be retro-actively slashed
+- See ⚠️  in `## Fees gathering` 
+
+❓ How to check balances with unsettled transactions ?
+- Fees are taken independently of transaction settlement. The fee part will be settled by the node itself.
+ 
+❓ How to take fees on blobs that can’t be prooved ?
+- Fees are taken independently of transaction settlement. The fee part will be settled by the node itself.
+
+❓ Fees are present in blob transaction, so they should pay for the proof transaction, which is the costly part. How can we anticipate the "cost" of the proof before having it, only with the blobs ?
+
+
+## Fees gathering
+
+Once the blob transaction is sequenced in a block, any node can generate the FeeProofTransaction for all fees blobs in the previous block(s), and once 
+this one is disseminated & included in a block, all fees will be effectively taken.
+```rust
+pub struct FeeProofTransaction {
+    pub transactions: Vec<TxHash>,
+    pub proof: Vec<u8>,
+}
+```
+
+⚠️  If a transaction was introduced in a block by a node that didn't check the identity. This FeeProofTransaction won't be able to be generated, (unless this single transaction is removed 
+from the `transactions` list, which makes the logic complexe where it shouldn't). That's why we need a guarantee that no transactions are sequenced if the identity is valid. To ensure that, all nodes needs to check `identity_proof` during data dissemination.
```

Comment:
C'est là que je me demande si on pourrait pas trigger un [IOU](https://en.wikipedia.org/wiki/IOU) mais sur le validateur qui a accepté la TX. Ou juste rétroactive slashing et accepter un mode "failure" de temps en temps.

Reply:
le souci, c'est qu'en générant la proof, il faudrait savoir quels sont les transactions "invalides", et générer une 2e preuve pour la même liste dont on a enlevé ces transactions, ça doit être possible en utilisant le program_outputs mais ça complexifie les chose

Reply:
qu'est-ce qui te dérange dans le fait de vérifier les identity proof avant de signer un batch dans la data dissemination ? ça fait beaucoup de proof à vérifier ?

Reply:
Mh en soit ça nous oblige soit à générer la proof client-side avec le blob (alors qu'une partie de l'intérêt c'était que le blob demande pas de proof) soit à s'assurer que le format du blob est valable (ce qui revient _un peu_ à une validation de signature, quoi que pas tout à fait)

Reply:
si je comprends bien, ton point est plus sur le fait d'avoir une identity proof dans la blob transaction que sur le fait de vérifier dans la data dissemination

donc si je te suis, il ne faudrait pas avoir d'identity proof dans la blob transaction et l'accepter sans plus de vérifications sur l'identité ? Pourquoi pas, dans ce cas on accepte toute blob transaction quand elle arrive (on ne peut faire aucune vérification sur le fait qu'on va pouvoir prendre des fees), et on attend la ProofTransaction qui contient la preuve d'identité. Et du coup pas d'anti spam grâce aux fees, c'est "gratuit" d'envoyer autant de blob qu'on veut, on ne paie que si on envoie un jour une proof transaction.

On va se retrouver à faire un "aws" gratuit... car on pourra envoyer des blobs sur la chaine sans frais

Reply:
A vrai dire j'ai pas une excellente réponse - je pense inclure la proof of identity à minima va souvent être nécessaire.

---

Le principe général est que les nodes sont tout à fait libres de ne pas inclure une transaction qui leur semble factice ou douteuse. Gérer les fees c'est un moyen d'éviter de se faire spam, mais qu'on a d'autres moyens de faire ça. En particulier, si les "blocks builders" ne sont plus des nodes dans le futur mais une auction à part par exemple, ça ne sera pas spécialement pertinent de le faire faire par les nodes.

L'autre partie de mon argument est qu'on a pas besoin de la `proof` pour vérifier l'identité, on a besoin de donnée. Dans le cas d'une identité validée par ECDSA, on peut juste envoyer la signature dans le blob et on peut vérifier avec du hardware natif comme sur ethereum dans la node.

Reply:
> qui leur semble factice ou douteuse

ça me semble compliqué de pas avoir un critère objectif et certain (une proof, une signature...)... autrement, si c'est sur des heuristiques, c'est toujours exploitable

> En particulier, si les "blocks builders" ne sont plus des nodes dans le futur mais une auction à part

que ce soit les nodes ou une auction à part je vois pas de différence, moi je vois "le réseau", s'il veut pas travailler trop gratuitement, il faut pouvoir prendre des fees, et donc le réseau vérifie qu'il va pouvoir prendre ces fees 

> pas besoin de la proof pour vérifier l'identité

oui, que ce soit zk proof ou signature c'est pareil, mais il faut bien une **garantie** que le `fees.payer` est "ok" pour payer les frais... il faut qu'on décide de comment on **garantie** une identité, et quand on en discutait il semblait qu'on voulait que tout soit zkProof.

Donc si je comprends ton idée, on partirait plutôt sur une V4 :

```rust
pub struct Blob {
    pub contract_name: ContractName,
    pub data: BlobData,
}

pub struct Fees {
    pub payer: Identity,
    pub amount: u64,
    pub fee_contract: ContractName,
    pub identity: Blob,
}

pub struct BlobTransaction {
    pub fees: Fees,
    pub identity: Identity,
    pub blobs: Vec<Blob>,
}
```

avec un `hydentity` qui prend une signature ECDSA en blob, signature qui peut être vérifiée par le réseau (aka la node pour l'heure) avant d'ajouter la transaction à une lane

Reply:
> que ce soit les nodes ou une auction à part je vois pas de différence, moi je vois "le réseau", s'il veut pas travailler trop gratuitement, il faut pouvoir prendre des fees, et donc le réseau vérifie qu'il va pouvoir prendre ces fees

Oui mais c'est pas forcément le même système et pas forcément notre responsabilité en tant que on écrit la node en ce moment. Mais bon ça reste conceptuel.

> ça me semble compliqué de pas avoir un critère objectif et certain (une proof, une signature...)... autrement, si c'est sur des heuristiques, c'est toujours exploitable
En soit oui, mais bon internet fonctionne :P 

---

Et en terme de ce que ça donne en fait ça serait ça pour moi:
```rs
 pub struct Fees {
     pub payer: Identity,
     pub fee: Blob,
     pub identity_blob: Blob,
     pub identity_proof: Option<Vec<u8>>
 }

 pub struct BlobTransaction {
     pub fees: Fees,
     pub identity: Identity,
     pub blobs: Vec<Blob>,
 }
```
Où  le node peut soit vérifier la proof si présente, soit tenter de parse le identity_blob. Et pour fee il faut parse le blob.